### PR TITLE
Support JWT key reload in pageserver

### LINKED ISSUE #4897: Support multiple keys for JWT validation
To make it possible to rotate keys, pageserver and safekeeper need to accept tokens signed by multiple keys.

### PR DESCRIPTION
**Pull request by arpad-m**
## Problem

For quickly rotating JWT secrets, we want to be able to reload the JWT public key file in the pageserver, and also support multiple JWT keys.

See #4897.

## Summary of changes

* Allow directories for the `auth_validation_public_key_path` config param instead of just files.
* Make the JWT public keys easily globally swappable by using the `arc-swap` crate.
* Add an endpoint to the pageserver, triggered by a POST to `/v1/reload_auth_validation_keys`, that reloads the JWT public keys from the pre-configured path (for security reasons, you cannot upload any keys yourself).

Fixes #4897

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by arpad-m** [comment]
I have added tests for key reloading now. There is also the multiple keys case which we could also maybe have tests for, idk.

File: test_runner/regress/test_auth.py [hunk 1]
```diff
@@ -82,6 +84,115 @@ def test_compute_auth_to_pageserver(neon_env_builder: NeonEnvBuilder):
...
+        Path(env.repo_dir) / "auth_public_key.pem" / "auth_public_key_old.pem"
+    )
+    os.rename(
+        Path(env.repo_dir) / "auth_public_key.pem", Path(env.repo_dir) / "auth_public_key.pem.file"
+    )
+    os.mkdir(Path(env.repo_dir) / "auth_public_key.pem")
+    os.rename(
+        Path(env.repo_dir) / "auth_public_key.pem.file",
+        Path(env.repo_dir) / "auth_public_key.pem" / "auth_public_key_old.pem",
+    )
```
Comment (hlinnaka) [inline test_runner/regress/test_auth.py:106]: It took me a long time to understand what this test does. It replaces the file `auth_public_key.pem`, with a _directory_ with the same name `auth_public_key.pem`. That's pretty confusing. Can we change the path in the pageserver config instead? At least add a comment to explain what this does.

File: test_runner/fixtures/neon_fixtures.py [hunk 2]
```diff
@@ -874,10 +875,34 @@ def get_pageserver_version(self) -> str:
 
     @cached_property
     def auth_keys(self) -> AuthKeys:
-        pub = (Path(self.repo_dir) / "auth_public_key.pem").read_text()
+        pem_path = self.auth_public_key_path or (Path(self.repo_dir) / "auth_public_key.pem")
+        pub = pem_path.read_text()
```
Comment (hlinnaka) [inline test_runner/fixtures/neon_fixtures.py:879]: Is `AuthKeys.pub` field actually used for anything?

File: test_runner/regress/test_auth.py [hunk 3]
```diff
@@ -82,6 +84,115 @@ def test_compute_auth_to_pageserver(neon_env_builder: NeonEnvBuilder):
...
+    env = neon_env_builder.init_start()
+
+    pageserver_token_old = env.auth_keys.generate_pageserver_token()
+    pageserver_http_client_old = env.pageserver.http_client(pageserver_token_old)
+
+    pageserver_http_client_old.reload_auth_validation_keys()
+
+    env.auth_public_key_path = (
+        Path(env.repo_dir) / "auth_public_key.pem" / "auth_public_key_old.pem"
+    )
```
Comment (hlinnaka) [inline test_runner/regress/test_auth.py:98]: You set this to match the path of the "old" key, but later when you regenerate the token, you don't change this path. So after regenerating the token, the AuthKeys has a mix of old public key path and new private key path.

I think that's harmless because the AuthKeys.pub path is not actually used for anything, but seems confusing.

File: test_runner/regress/test_auth.py [hunk 4]
```diff
@@ -82,6 +84,115 @@ def test_compute_auth_to_pageserver(neon_env_builder: NeonEnvBuilder):
...
+    # Add a new key pair
+    env.regenerate_keys_at(
+        Path("auth_private_key.pem"), Path("auth_public_key.pem/auth_public_key_new.pem")
+    )
+
+    # Reload the keys on the pageserver side
+    pageserver_http_client_old.reload_auth_validation_keys()
+
+    # We can continue using the old token
+    pageserver_http_client_old.reload_auth_validation_keys()
```
Comment (hlinnaka) [inline test_runner/regress/test_auth.py:117]: Is this duplicated on purpose?

File: libs/utils/src/auth.rs [hunk 5]
```diff
@@ -47,31 +48,75 @@ impl Claims {
...
+        for decoding_key in &self.decoding_keys {
+            res = Some(decode(token, decoding_key, &self.validation));
+            if let Some(Ok(res)) = res {
+                return Ok(res);
+            }
+        }
+        if let Some(res) = res {
+            res.map_err(anyhow::Error::new)
+        } else {
+            anyhow::bail!("no JWT token authorized")
```
Comment (hlinnaka) [inline libs/utils/src/auth.rs:118]:
```suggestion
            anyhow::bail!("no JWT decoding keys configured")
```
If I understand this correctly, you end up here if `decoding_keys` is empty.

File: libs/utils/src/auth.rs [hunk 6]
```diff
@@ -47,31 +48,75 @@ impl Claims {
...
-        Ok(decode(token, &self.decoding_key, &self.validation)?)
+        let mut res = None;
+        for decoding_key in &self.decoding_keys {
+            res = Some(decode(token, decoding_key, &self.validation));
+            if let Some(Ok(res)) = res {
+                return Ok(res);
+            }
+        }
+        if let Some(res) = res {
+            res.map_err(anyhow::Error::new)
```
Comment (hlinnaka) [inline libs/utils/src/auth.rs:116]:
```suggestion
            // return the successful result, or last error if none of the keys worked.
            res.map_err(anyhow::Error::new)
```

File: libs/utils/src/auth.rs (hunk 5 above)
Comment (hlinnaka) [reply libs/utils/src/auth.rs:118]: I wonder if we should warn or error out earlier, if authentication is enabled but the directory is empty

File: test_runner/regress/test_auth.py (hunk 4 above)
Comment (arpad-m) [reply test_runner/regress/test_auth.py:117]: yes, it's there to ensure that authenticated http requests to the pageserver work (of which the reload function is one). maybe I should factor this out into a function to make it more clear.

File: test_runner/regress/test_auth.py (hunk 1 above)
Comment (arpad-m) [reply test_runner/regress/test_auth.py:106]: I originally explored customizing the path, but it is handed through a lot of layers (including the one in the `control_plane` directory) until it finally lands in the pageserver's config. I thought that doing this many changes just to have a nicer directory name isn't that useful.

Good point about it being confusing, I will add a comment and see what other reviewers think about it.

File: libs/utils/src/auth.rs (hunk 6 above)
Comment (arpad-m) [reply libs/utils/src/auth.rs:116]: I have added a rustdoc comment to the function.

File: libs/utils/src/auth.rs (hunk 5 above)
Comment (arpad-m) [reply libs/utils/src/auth.rs]: I've added a commit that adds warnings upon load.

File: test_runner/regress/test_auth.py (hunk 4 above)
Comment (arpad-m) [reply test_runner/regress/test_auth.py]: I've moved it into a new function in a subsequent commit.

File: test_runner/regress/test_auth.py (hunk 3 above)
Comment (arpad-m) [reply test_runner/regress/test_auth.py]: I've removed the public key part of `AuthKeys`.

File: test_runner/fixtures/neon_fixtures.py (hunk 2 above)
Comment (arpad-m) [reply test_runner/fixtures/neon_fixtures.py:879]: I've removed the public key part of `AuthKeys`.

File: test_runner/regress/test_auth.py (hunk 1 above)
Comment (arpad-m) [reply test_runner/regress/test_auth.py]: Comment added

File: libs/utils/src/auth.rs [hunk 7]
```diff
@@ -47,31 +49,83 @@ impl Claims {
...
+                .collect::<Result<Vec<_>>>()?
+        } else if metadata.is_file() {
+            let public_key = fs::read(key_path)?;
+            vec![DecodingKey::from_ed_pem(&public_key)?]
+        } else {
+            anyhow::bail!("path is neither a directory or a file")
+        };
+        if decoding_keys.is_empty() {
+            warn!("Configured for JWT auth with zero decoding keys. All JWT gated requests will be rejected.");
+        }
```
Comment (koivunej) [inline libs/utils/src/auth.rs:107]: Shouldn't we just error instead? If this happens during startup, we can never be ordered to reload keys, and shutting down pageserver is the only option?

Comment (koivunej) [reply libs/utils/src/auth.rs:107]: I think it would be in line with the above "all keys need to be loadadble".

File: pageserver/src/http/routes.rs [hunk 8]
```diff
@@ -353,6 +354,32 @@ async fn status_handler(
...
+    let key_path = config.auth_validation_public_key_path.as_ref().unwrap();
+    info!("Reloading public key(s) for verifying JWT tokens from {key_path:#?}");
+
+    match JwtAuth::from_key_path(key_path) {
+        Ok(new_auth) => {
+            shared_auth.swap(new_auth);
+            json_response(StatusCode::OK, ())
+        }
+        Err(e) => {
+            warn!("Error reloading public keys from {key_path:#?}: {e:}");
```
Comment (koivunej) [inline pageserver/src/http/routes.rs:377]: Does the `:#?` have some special meaning with camino paths?

Comment (arpad-m) [reply pageserver/src/http/routes.rs:377]: I copy pasted it from the `"Loading public key(s) for verifying JWT tokens from {key_path:#?}"`. The `:#?` [indicates](https://doc.rust-lang.org/nightly/std/fmt/index.html#sign0
) "alternate" debug formatting which generally means extra line breaks and indentation. In this instance, it might actually not make a difference at all, as through many layers the path `Debug` impl boils down to:

https://github.com/rust-lang/rust/blob/698db856de0b67313ddcb96b6599598058489ea9/library/core/src/str/lossy.rs#L78-L108

You don't see the [alternate function](https://doc.rust-lang.org/nightly/std/fmt/struct.Formatter.html#method.alternate) being called to query if we are in alternate mode.

In other words, we can probably remove it.

File: libs/utils/src/auth.rs (hunk 7 above)
Comment (arpad-m) [reply libs/utils/src/auth.rs:107]: Made it bail instead.

File: pageserver/src/http/routes.rs (hunk 8 above)
Comment (arpad-m) [reply pageserver/src/http/routes.rs]: Removed the `:#?` stuff and replaced it with `:?`

Comment (koivunej) [reply pageserver/src/http/routes.rs]: Yep, the `{:?}` is fine if you want it quoted, which works great here. Before asking I did do some testing but failed to see any difference.

> core/str/lossy.rs

Hmm I thought this was a camino path but got even confused.

Resolving.

File: libs/utils/src/auth.rs (hunk 7 above)
Comment (koivunej) [reply libs/utils/src/auth.rs]: Did you forget to push?

**Review by koivunej — APPROVED** [review]
The tests reusing the original key file path as a directory is a bit confusing but I think it's more ergonomic than to fiddle with the config generation, and it shows we are not sticky with "this used to be a file but it is now a directory, bail!".

I still want that failure in pageserver startup if we fail to read any keys though (~I think it was left unpushed~ [c425c1d](https://github.com/neondatabase/neon/pull/5594/commits/c425c1dc68065c0c67bbbb4cb8479d2e6b4c99a6)).

File: libs/utils/src/auth.rs [hunk 9]
```diff
@@ -47,31 +48,83 @@ impl Claims {
...
+                })
+                .collect::<Result<Vec<_>>>()?
+        } else if metadata.is_file() {
+            let public_key = fs::read(key_path)?;
+            vec![DecodingKey::from_ed_pem(&public_key)?]
+        } else {
+            anyhow::bail!("path is neither a directory or a file")
+        };
+        if decoding_keys.is_empty() {
+            anyhow::bail!("Configured for JWT auth with zero decoding keys. All JWT gated requests will be rejected.");
```
Comment (koivunej) [inline libs/utils/src/auth.rs:105]:
```suggestion
            anyhow::bail!("Configured for JWT auth with zero decoding keys. All JWT gated requests would be rejected.");
```

File: libs/utils/src/auth.rs (hunk 7 above)
Comment (koivunej) [reply libs/utils/src/auth.rs]: Oki it is there, I must've been looking at an outdated diff.

**Review by petuhovskiy — APPROVED** [review]

File: pageserver/src/http/routes.rs [hunk 10]
```diff
@@ -354,6 +355,32 @@ async fn status_handler(
     json_response(StatusCode::OK, StatusResponse { id: config.id })
 }
 
+async fn reload_auth_validation_keys_handler(
```
[the remark this answers is not in the document]
Comment (arpad-m) [reply pageserver/src/http/routes.rs:358]: > Can we just restart pageserver?

We can, but it will mean downtime, at least before HA work. With HA work, it will not be directly responsible for downtime, but it will reduce the number of pageservers that can respond so it will degrade our service quality for the time of the key rollout. Such a degradation would already be a regular occurence due to deploys so one could say that it's an okay cost. Then #4897 would not be needed.

>  then it looks like a more generic 'live-reload pageserver configuration'

A generic live reload feature was one of the alternatives I laid out [here](https://github.com/neondatabase/neon/issues/4897#issuecomment-1668597469), but I agree with John that something specific just for reloading the specified jwt path (directory or file) is better because it's easier to implement.

Note that this PR does not allow a change of the actual parts stored in toml, namely the path. That path remains the same, the only thing that is reloaded is the config that the path points to (as we don't put the jwt keys into the toml file directly).

The generic ability for the pageserver to reload arbitrary configuration is a big feature. We probably would only support a subset of config params, and would have to start a discussion about what that subset should be, probably influenced by the use cases. To me it feels that there is few such use cases where, but I might be wrong.

Right now we *need* only the jwt tokens. And this live reload ability is just a clone away from being broken (at least for easily cloneable types like ints, paths or strings. People might do such clones without noticing, so I wonder how we should preserve the live reload ability after we add it. 

The other issue is how to manage the ability to change the config from a central point and at the same time, be able to access it from multiple threads.

>  having a specific http API just to reload keys, looks a bit strange to me, but maybe I miss something.

it is easier to implement on the pageserver side, as we already have multiple http apis, compared to SIGHUP. But I'm fine with SIGHUP as well. I can look into it.

[the remark this answers is not in the document]
Comment (ololobus) [reply pageserver/src/http/routes.rs:358]: > it is easier to implement on the pageserver side, as we already have multiple http apis, compared to SIGHUP. But I'm fine with SIGHUP as well. I can look into it.

No, I'm fine with http API too if generic live-reload is more complicated and you wanted to narrow down the scope. Thanks for the explanation

I still have two comments

1. It'd be nice to add an OpenAPI spec for this new API, so it was clear how to use

2. Shouldn't it be just empty-body `POST /reload_auth_validation_keys`? If path is set in the toml config, then ps has to only re-read the keys directory, right? From the python test it looks like that, but in Rust code it's read from the request?

[the remark this answers is not in the document]
Comment (arpad-m) [reply pageserver/src/http/routes.rs:358]: > It'd be nice to add an OpenAPI spec for this new API

Good point, I will.

> If path is set in the toml config, then ps has to only re-read the keys directory, right?

Yes, precisely. The API call only does the reload, but nothing else. You cannot specify any directory or such, and I've been advised against adding the ability to specify a directory by @ tychoish.

[the remark this answers is not in the document]
Comment (nikitakalyanov) [reply pageserver/src/http/routes.rs:358]: So, we'll have to update the deploy scripts to run this API?

File: safekeeper/src/lib.rs [hunk 11]
```diff
@@ -69,7 +72,7 @@ pub struct SafeKeeperConf {
     pub wal_backup_enabled: bool,
     pub pg_auth: Option<Arc<JwtAuth>>,
     pub pg_tenant_only_auth: Option<Arc<JwtAuth>>,
```
Comment (nikitakalyanov) [inline safekeeper/src/lib.rs:74]: Hmm, why not put SwappableJwtAuth for the other params too? (pg_tenant_only_auth, pg_auth)

File: test_runner/regress/test_auth.py [hunk 12]
```diff
@@ -82,6 +91,94 @@ def test_compute_auth_to_pageserver(neon_env_builder: NeonEnvBuilder):
...
 
+def test_pageserver_multiple_keys(neon_env_builder: NeonEnvBuilder):
+    neon_env_builder.auth_enabled = True
+    env = neon_env_builder.init_start()
+    env.pageserver.allowed_errors.append(".*Unauthorized: malformed jwt token.*")
+
+    pageserver_token_old = env.auth_keys.generate_pageserver_token()
+    pageserver_http_client_old = env.pageserver.http_client(pageserver_token_old)
+
+    pageserver_http_client_old.reload_auth_validation_keys()
```
Comment (nikitakalyanov) [inline test_runner/regress/test_auth.py:102]: Why is this first reload needed?

File: libs/utils/src/auth.rs [hunk 13]
```diff
@@ -47,31 +48,83 @@ impl Claims {
...
         }
     }
 
     pub fn from_key_path(key_path: &Utf8Path) -> Result<Self> {
-        let public_key = fs::read(key_path)?;
-        Ok(Self::new(DecodingKey::from_ed_pem(&public_key)?))
+        let metadata = key_path.metadata()?;
+        let decoding_keys = if metadata.is_dir() {
+            fs::read_dir(key_path)?
+                .map(|entry| {
```
Comment (nikitakalyanov) [inline libs/utils/src/auth.rs:93]: don't we need to filter only files here so that nested directories don't break us?

**Review by nikitakalyanov — APPROVED** [review]
LGTM
I am not sure the HTTP API will be convenient. Do we just put it in an ansible handler and call each deploy? But it seems useful at least for tests which is nice

I also see that this PR touches safekeepers too. So let's update the title and make the other 2 safekeeper params work

File: test_runner/regress/test_auth.py (hunk 12 above)
Comment (arpad-m) [reply test_runner/regress/test_auth.py:102]: it's there to ensure that a) reloading works when nothing has changed and b) that jwt gated http calls work. it's not needed for any actual key swapping routine in production or such.

File: libs/utils/src/auth.rs (hunk 13 above)
Comment (arpad-m) [reply libs/utils/src/auth.rs:93]: it's not needed technically (we can require that the directory only contains the key files and nothing else), but I will make a change to just ignore these directories.

File: safekeeper/src/lib.rs (hunk 11 above)
Comment (arpad-m) [reply safekeeper/src/lib.rs:74]: I made this PR only for the pageserver. the safekeepers need more work than just putting `SwappableJwtAuth` to those items, they also need that HTTP endpoint for reloading for example. I did change the safekeepers because the safekeepers use some code that is shared with the pageserver.

**Comment by arpad-m** [comment]
> I am not sure the HTTP API will be convenient.

@nikitakalyanov what would you suggest for triggering the live-reload instead of the HTTP API?

**Comment by nikitakalyanov** [comment]
> what would you suggest for triggering the live-reload instead of the HTTP API?

I think practically we will just wait for the next deploy to trigger a restart and reload the keys. Because otherwise, someone would have to write some ansible logic to detect when calling this API is necessary. That sounds error-prone. I am not against the reload API because it helps in testing. I just think that it's unlikely to be used during deploys. Maybe only manually.

**Comment by arpad-m** [comment]
>  I think practically we will just wait for the next deploy to trigger a restart and reload the keys. Because otherwise, someone would have to write some ansible logic to detect when calling this API is necessary.

Ideally if some JWT key gets leaked we should be able to quickly deploy new keys (without an actual deploy that has PS restarts and temporary service degradation due to the restart). But this doesn't have to be implemented at the start.

File: pageserver/src/http/routes.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (arpad-m) [reply pageserver/src/http/routes.rs:396]: > we'll have to update the deploy scripts to run this API?

if you want to call the API, sure. but it's not obligatory, only if we want to support key rollover without needing restarts (which is a good idea IMO).
